src/libsyntax_ext/proc_macro_harness.rs

```diff
@@ -360,6 +374,10 @@ fn mk_decls(
             proc_macro, bridge, client, proc_macro_ty, method,
         ]));
         custom_derives.iter().map(|cd| {
+            proc_macro_infos.borrow_mut().push(ProcMacroInfo {
+                span: cd.span,
+                attrs: cd.raw_attrs.clone()
+            });
```

Comment:
@petrochenkov I guess this is what I want to make sure you have nothing against.

Reply:
This is something I have a lot against, actually.

Aren't `fn` items encoded into the metadata already (including spans and attributes)?
As I understand it, a compiled proc macro crate has two views - the "regular" one with the original `fn` items and everything, and the "proc-macro facade" one masking all the original items and making it look like the crate only exposes macros.

So we just need to link from the proc-macro view (`CrateMetadata::proc_macros`) to the real items somehow instead of cloning some data from those items, keeping them in AST and encoding them for the second time.

Reply:
Ah yeah, we'd just need to map the DefId's.

Reply:
Would it make sense to unconditionally decode and store the 'real' `def_path_table`, but continue to use a 'fake' table for most operations on proc macro crates?

Reply:
I suppose, yeah. At this point I'm not sure why we don't use the original `DefId`s directly, and just change what `def_kind` returns?